File: 46.全排列.py
#
# @lc app=leetcode.cn id=46 lang=python3
#
# [46] 全排列
#
from typing import List

# @lc code=start
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        self.backtrack(res, [], nums)
        return res

    def backtrack(self, res, path, candidates):
        if len(candidates) == 0:
            res.append(path[:])
            return
        for i, num in enumerate(candidates):
            path.append(num)

            # ignore candidates[i] and use the other elements
            self.backtrack(res, path, candidates[:i]+candidates[i+1:])
            
            path.pop()
    
# @lc code=end
s = Solution()
print(s.permute([1,2,3,4]))

File: 46.全排列.java
import java.util.List;
import java.util.ArrayList;
import java.util.Stack;

/*
 * @lc app=leetcode.cn id=46 lang=java
 *
 * [46] 全排列
 */

// @lc code=start
class Solution {
    List<List<Integer>> res;

    public List<List<Integer>> permute(int[] nums) {
        res = new ArrayList<>();
        Stack<Integer> track = new Stack<>();
        backtrack(nums, track);
        return res;
    }

    private void backtrack(int[] nums, Stack<Integer> track) {
        if (track.size() == nums.length) {
            res.add(new ArrayList<Integer>(track));
            return;
        }
        for (int i = 0; i < nums.length; ++i) {
            if (track.contains(nums[i]))  continue;
            track.add(nums[i]);
            backtrack(nums, track);
            track.pop();
        }
    }
}
// @lc code=end

